### DISCUSSION TITLE
Overall architecture

### DISCUSSION THREAD

**Discussion by copych** [copych/AciduinoBox#5] [Ideas]
@rdridder @midilab
Now that the code compiles, let's discuss what device we are constructing.
First question for me is "will the sequencer only transmit noteOn/noteOff events, or will there be also some CC automation?" If we choose the first option, it would be closer to the straight forward 303/808 concept, while the second option would be slightly more advanced, and would add some sugar, like standard control ramps, rec/play automations etc., but this requires more coding. Also the first approach means a lot of pots, knobs and so on, while the second could do the thing with less pots, but some well-thinked GUI.
Please, share your thoughts.
@midilab please, specify what is the best place for setting callbacks for midi events, as we don't want to really transmit them through the wires to the synths.

**Comment by rdridder** [comment]
@copych @midilab
I am going to submit another pull request which has some pin remappings. I now have the standard layout working that was documented by @midilab, see: [ESP32 hardware](https://github.com/midilab/aciduino/blob/master/v2/hardware/imgs/aciduino_v2-esp32_bb.png). In the image one button is not connected, I guess that should be the transport button. 

I have connected the button in my hardware version. See image below. This hardware corresponds with the pull request I will submit shortly and it corresponds to the midilab layout with the additional button connected. I think we should all start with this hardware and built from there.

![basic-layout](https://github.com/user-attachments/assets/b614e350-f618-47d5-91b9-d2b8a2ca55d6)

In the pull request I also hardcoded the timer frequency for uClock, wit hthe default code it did not run. We have to dig in the specifics. I calculated those frequencies a while back but I need to dig into it again to make sure everything is correct.

uCtrl contains a lot of options to use pots, encoders and buttons, so integrating the hardware should be no problem if AciduinoBox is running stable with the minimum setup. My two cents would be to create the default layout, make the basics work on that and then take next steps. We need the baseline hardware and the fully working software. If the basics are all working nicely together we can then expand upon the stable base.

To get the software working we need to dive into the specifics of the timers and tie the tasks used by those timers to core1. Another big topic is the integration between the synths/drummachines and the sequencer. My gut feeling says we need to utilize uCtrl to tie them together. Probably we need to alter the midi module or we can add an additional module. @midilab can probably advice her.

I currently have a problem with the [Lolin S3](https://www.wemos.cc/en/latest/s3/s3.html). I have some Acidbox PCB's but I don't get any sound from the Lolin. I see the DAC is powered on and I turned on the debug output but I get no sound. I also have some cheap ESP32 dev boards and they actually produce sound. So I would need to sort that out. @copych, do you have any suggestions or advice ? The [Acidbox PCB](https://github.com/streetuff/AcidBox-PCB) I used also uses a Lolin S3 so I gues there should be a way to make it work.

**Comment by copych** [comment]
@rdridder thank you for the explanations and your detailed point of view. I mostly agree, so let's do it. @midilab 's help would be highly appreciated, otherwise we're to make unneeded inventions.
About Lolin S3 and ESP32-S3 in general. Pinouts of the two models (ESP32 and ESP32-S3) are quite different despite the similar naming. So please don't mix them. I never used PCBs, they are not of my authority, so I'd recommend to test a simple breadboard prototype with a minimal wiring and JUKEBOX defined. Default pins for I2S are 5, 6 and 7. 
https://github.com/copych/AcidBox/blob/main/hardware/2023-03-28%20at%2009.19.49.jpeg

**Reply by rdridder** [reply]
Hi @copych,

Thank you for your suggestions. I tried the Lolin and some other S3 board I had. The generic S3 board immediatly started playing the Acidbange, the Lolin board does not want to cooperate. I will check some more because it seems there is a difference between the two boards.

**Reply by rdridder** [reply]
I finally found the problem. I started a long time ago with my original setup and I had to connect some islands at the back of the DAC chip with solder. I did this with the original DAC but not with the additional ones I had lying around. So basically a matter of configuring the DAC and it works !

**Reply by copych** [reply]
Great to know that )

**Comment by rdridder** [comment]
I am playing around a bit with draw.io to create a suitable hardware layout. For now it is basically ideas but it might be good to get the conversation started. The diagram is in my fork in draw.io folder, here is the first sketch, obviously incomplete to drive the whole UI but I think the contextual pots might be a good idea to use.

![aciduinobox-Page-1](https://github.com/user-attachments/assets/7892cb9b-2599-4906-afa2-bdf56f17e72c)

**Reply by copych** [reply]
Looks nice! I imagined something like this with the contextual pots or encoders. 
My personal preferences would be a track selector at the left (just three buttons, allowing simultaneous selecting, so that, say, a pot could control two tracks at a time). Track mute button is a must IMHO. Also dedicated controls for distortion and accent level. The question for me is contextual pots vs contextual encoders... Both have pros and cons, e.g. encoders require twice as many GPIOs, but pots would require some "hacks" to be shared between different tracks or controls. Also pots are noisy by nature, and getting worse while wearing out... Things to think of and to discuss.

**Reply by copych** [reply]
Thank you for the idea with draw.io, never knew that such tool exists. Here's what I was thinking of

![image](https://github.com/user-attachments/assets/75ae9493-d8c2-4eef-9eda-ce5a9cb92270)

Though I don't know if it's possible at all to run such many encoders :-/

30 buttons could be a matrix of 5x6 = 11 GPIOs. Encoder for param input is 3 GPIOs. 6 pots or 6 encoders = 6 or 12 GPIOs, better not muxed.

**Reply by rdridder** [reply]
That looks like a really nice layout, great work.

I found this thread on the Teensy forum: https://forum.pjrc.com/index.php?threads/multiplexing-rotary-encoders-or-using-port-expander-what-will-work-with-the-library.59485/

They are using a clever technique to use a simple shift register. If we use a shift register as suggested we could use multiple encoders with minimal wiring. I followed the link in the thread but that library is obsolete, this is the current library: https://github.com/luni64/EncoderTool

For the leds we could use addressable leds like the WS2812B. All the buttons could be multiplexed the same way as the encoders.

My suggestion would be to try out the multiplexing setup with the CD4067 breakout boards or with the 74HC165 shift register. I have some encoders laying around but I am not sure if I have any shift registers. I would need to check what I have. If all works as expected we can even add some more encoders for some 808 parameters and the effects. I will do some digging through my components and let you know if I find something. Can you also check your components or if you don't have anything laying around settle on a specific shift register to make the software more easily testable ?

Forgot to say but I like encoders more than pots, so I am 100% on board with the encoder approach.

**Reply by copych** [reply]
I've got a couple of 4067's, and encoders of 2 types. Some Ws2812b as well. Also I have ordered a couple of i2c pcf8575 port extenders, they arrive in a couple of weeks.. No shift registers though, and i couldn't find breakout boards for 74hc165. Glad that using encoders is accepted that easy )

**Reply by copych** [reply]
4067 is an analog multiplexer with good timing specs, but to get 16 muxed lines you still have to give 5 GPIOs, but if they say that polling is ok for muxed lines, then we can try. Probably we should run a high priority task just for polling controls on Core1 along with the UI and sequencer task.
And probably we'll have to write the custom polling routine to combine button matrix and encoders connected thru 2 multiplexers, cause matrix won't work on a single multiplexer if I understand it right.
Also if I remember correctly, for simultaneously pushed buttons we'll have to use diodes in matrix.

**Reply by rdridder** [reply]
I think we just have to try out if polling suits our needs but I guess it should not be a problem. We can also run a timer that polls each x amount of milliseconds. I haven't checked out the code in the library in detail but in the examples they are basically running a tick in the loop() and then checking in a loop if a value changed, seems simple enough. We could also adopt that approach.

I suggest we start with the encoders to extend the basic layout and take it from there. Just adding one type of component at a time. A good second candidate will probably be the buttons.

**Reply by rdridder** [reply]
I have checked my component stash but I don't have any 4067 muxes. I have ordered some breakout boards which will probably arrive next week. In the meantime I will check if I can extend the midi signals a bit so the synths can also receive midi controls like cutoff/resonance etc. I was thinking about your layout and it is actually very clever to have the select buttons to select a synth. This will make it possible to either feed a midi control to either one of the synths or both. In combination with an encoder this will yield a very nice way to control the parameters since an encoder only sends ups and downs instead of an absolute value. Otherwise we would experience jumps when setting a high cutoff on synth1 and then switch to synth2 with a low cutoff.

**Comment by copych** [comment]
> Otherwise we would experience jumps when setting a high cutoff on synth1 and then switch to synth2 with a low cutoff.

This is what I meant saying "we'll have to use some hacks".

I have ordered these https://aliexpress.com/item/1005006077226247.html?sku_id=12000035627199590&srcSns=sns_WhatsApp&businessType=ProductDetail&spreadType=socialShare&tt=MG&utm_medium=sharing encoders in addition to what I have for now. As I already have a couple of 4067's, I can build a stand to test polling algorithms along with button matrix. And according to what I have read, it's confirmed that we need diodes to safely process simultaneous connecting of several pins.

**Reply by rdridder** [reply]
Ah nice. I have a bunch of those [encoder breakout boards](https://nl.aliexpress.com/item/1005006551162496.html) which will be fine for the testing phase.

I would also check out that library I linked. Here they are using 6 pins to process 16 encoders: https://github.com/luni64/EncoderTool/blob/master/examples/2_multiplexing/multiplexed_4067/multiplexed_4067.ino
Basically using the dual 4067 setup.

**Reply by copych** [reply]
I wouldn't like to have 4 of 4067's, as they will occupy 20 GPIOs. so maybe we can arrange everything in this manner?
![image](https://github.com/user-attachments/assets/ce2946cc-14cd-42ad-979c-b8010cca5011)

**Reply by rdridder** [reply]
> I wouldn't like to have 4 of 4067's, as they will occupy 20 GPIOs. so maybe we can arrange everything in this manner?

If you have 4 4067 they will only occupy 8 IOS's. You tie S0, S1, S2 and S3 together. When you send the state to the mux you write to S0,S1,S2,S3 and because the S ports are tied together all 4 muxes will be in the same state. Now you can read out all 4 muxes via the data pins. So basically what is needed is S0-S3 and 4 data pins.

Please see example schematic below (Please note, it is a bit simplified but the nets show the needed pins). I have drawn 3 muxes, this facilitates 8 encoders and 24 switches.

![Schematic_AciduinoBox_2024-12-11](https://github.com/user-attachments/assets/106c9e77-c294-4fe4-8b7c-841c93450c3d)

So if we want to read C0 we need to put S0 low, S1 low, S2 low and S3 low. Now we can read C0 for all three muxes. If we want to read C1 we want to set S0 high, S1 low, S2 low and S3 low. Now we can read C1 for all three muxes, etc. So pin wise using the muxes would be more efficient than using a matrix.

**Reply by copych** [reply]
Thanks, I now have seen stack-soldered breakout boards and got the idea of synchronizing S-lines. So with 4 muxes we'll need 4 GPIO's for S-lines and 4 for SIG-lines, and it'll give us 48 C-lines which we can use in any combination of buttons and encoders, and no additional diodes/resistors and switching INPUT/OUTPUT pin modes.

**Reply by midilab** [reply]
I strongly suggest to go with uCtrl builtin drivers for pots and encoders, since aciduino is already coded for up to 16 pots via 4051 or 4067 and encoder support via 165 shiftregister(using multiplexer for digital encoders is not a good choice, shiftregisters are made to work with digital signal instead of mlutiplexer that handles analog signals).

Please take a time to read acudino and uCtrl documentation.

You can use 1 shared SPI for digital input and output via shiftregisters(buttons/encoders/leds).

**Reply by rdridder** [reply]
> using multiplexer for digital encoders is not a good choice, shiftregisters are made to work with digital signal instead of multiplexer that handles analog signals

Can you elaborate a bit more on this ? As far as I understand an analog MUX can be used for analog and digital signals. I tried to research the differences between analog and digital MUX'es and in short it looks like a digital MUX has a buffer to recondition/amplify the signal and an analog MUX just switches.

I am going to read the documentation for uCtrl and the encoder approach because it would definitely have preference to use the framework. However, I don't want to write off the 4067 approach entirely because we could also integrate a solution based on 4067 as a uCtrl module.

Thank you for your suggestions !

**Reply by midilab** [reply]
> Can you elaborate a bit more on this ? As far as I understand an analog MUX can be used for analog and digital signals. I tried to research the differences between analog and digital MUX'es and in short it looks like a digital MUX has a buffer to recondition/amplify the signal and an analog MUX just switches.

Indeed Muxes can be used for digital signal, but a shiftregister was made for digital signals, you can communicate faster than a mux, at the speed of a SPI interface and have a way more robust and clear signal to read, specialy when you want to read a lots of encoders attached to a 165(4 encoders per CI). the port switch of mux will add delay to change port while shiftregister is one bit in, one bit out at high speed. Its a matter of use the rigth tool for the job.

**Comment by copych** [comment]
@midilab Thank you for your suggestion. Please, recommend, what hardware we should use for such controls layout to work with the built-in drivers?
 
![image](https://github.com/user-attachments/assets/584e0413-424e-4a44-bf1f-2a0118dba969)

**Reply by midilab** [reply]
If you decide to go with encoders for synth controls you'll need to code support for it by your self, i do recomend to keep pots for synth controls since aciduino already supports up to 16 pots for direct track parameters controls and control learn function.

this is one of mine aciduinos with 16 pots.

![ctrl16](https://github.com/user-attachments/assets/cb3cd900-6f9b-45eb-9077-2ff5e66d76fd)

**Reply by rdridder** [reply]
We would like to use encoders to prevent bumps when switching context from one 303 to the other. So I guess we will experiment a bit what the best approach will be to use encoders.

**Reply by midilab** [reply]
uCtrl analog inputs were coded to avoid those bumps via software. I Suggest try it first the aciduino pots extension.
